internal/extender/overhead.go

```diff
@@ -39,151 +37,165 @@ var (
+	}
+	return nodeRequests
 }
 
-// GetOverhead fills overhead information for given nodes, and falls back to the median overhead
-// of the instance group if the node is not found
+// GetOverhead fills overhead information for given nodes.
 func (o OverheadComputer) GetOverhead(ctx context.Context, nodes []*v1.Node) resources.NodeGroupResources {
-	o.overheadLock.RLock()
-	defer o.overheadLock.RUnlock()
-	return o.getOverheadByNode(ctx, o.latestOverhead, nodes)
+	ov, _ := o.getOverheadByNode(ctx, nodes)
+	return ov
 }
 
-// GetNonSchedulableOverhead fills non-schedulable overhead information for given nodes, and falls back to the median overhead
-// of the instance group if the node is not found.
+// GetNonSchedulableOverhead fills non-schedulable overhead information for given nodes.
 // Non-schedulable overhead is overhead by pods that are running, but do not have 'spark-scheduler' as their scheduler name.
 func (o OverheadComputer) GetNonSchedulableOverhead(ctx context.Context, nodes []*v1.Node) resources.NodeGroupResources {
-	o.overheadLock.RLock()
-	defer o.overheadLock.RUnlock()
-	return o.getOverheadByNode(ctx, o.latestNonSchedulableOverhead, nodes)
+	_, nso := o.getOverheadByNode(ctx, nodes)
+	return nso
 }
 
-func (o *OverheadComputer) doStart(ctx context.Context) error {
-	t := time.NewTicker(30 * time.Second)
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case <-t.C:
-			o.compute(ctx)
-		}
+// getOverheadByNode computes and returns the overhead per node name.
+// This returns (overhead per node, nonSchedulableOverhead per node).
+func (o OverheadComputer) getOverheadByNode(ctx context.Context, nodes []*v1.Node) (resources.NodeGroupResources, resources.NodeGroupResources) {
+	overhead := resources.NodeGroupResources{}
+	nonSchedulableOverhead := resources.NodeGroupResources{}
+
+	for _, n := range nodes {
+		ov, nso := o.computeNodeOverhead(ctx, n.Name)
+		overhead[n.Name] = ov
+		nonSchedulableOverhead[n.Name] = nso
 	}
+	svc1log.FromContext(ctx).Debug("using overhead for nodes", svc1log.SafeParam("overhead", overhead), svc1log.SafeParam("nonSchedulableOverhead", nonSchedulableOverhead))
+	return overhead, nonSchedulableOverhead
 }
 
-func (o *OverheadComputer) compute(ctx context.Context) {
-	pods, err := o.podLister.List(labels.Everything())
-	if err != nil {
-		svc1log.FromContext(ctx).Error("failed to list pods", svc1log.Stacktrace(err))
-		return
-	}
-	rrs := o.resourceReservations.List()
-	podsWithRRs := make(map[string]bool, len(rrs))
-	for _, rr := range rrs {
-		for _, podName := range rr.Status.Pods {
-			podsWithRRs[podName] = true
-		}
+// computeNodeOverhead adds the requests of pods that don't have reservations and are counted as overhead.
+// This returns (overhead, nonSchedulableOverhead).
+func (o *OverheadComputer) computeNodeOverhead(ctx context.Context, nodeName string) (*resources.Resources, *resources.Resources) {
+	o.overheadLock.RLock()
+	defer o.overheadLock.RUnlock()
```

Comment:
Pushed this down here since we don't have pod movements across nodes, but if we think this might start conflicting too much with the informer updates and slow down scheduling, we can also lock once when getting overhead for all nodes.